crates/flv/src/script.rs

```diff
@@ -1,35 +1,246 @@
-    /// The name of the script data
-    pub name: String,
-    /// The data of the script data
-    pub data: Vec<Amf0Value<'static>>,
+pub struct OnXmpData {
+    live_xml: Option<String>,
+    other: HashMap<String, Amf0Value<'static>>,
+}
+
+impl TryFrom<Amf0Object<'_>> for OnXmpData {
+    type Error = Error;
+
+    fn try_from(value: Amf0Object<'_>) -> Result<Self, Self::Error> {
+        let mut other = HashMap::new();
+
+        let mut live_xml = None;
+
+        for (key, value) in value.iter() {
+            if key == "liveXML" {
+                live_xml = Some(value.as_string()?.to_string());
+            } else {
+                other.insert(key.to_string(), value.to_owned());
+            }
+        }
+
+        Ok(Self { live_xml, other })
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum ScriptData {
+    // Boxed because it's so big
+    OnMetaData(Box<OnMetaData>),
+    OnXmpData(OnXmpData),
+    Other {
+        /// The name of the script data
+        name: String,
+        /// The data of the script data
+        data: Vec<Amf0Value<'static>>,
+    },
 }
 
 impl ScriptData {
-    pub fn demux(reader: &mut io::Cursor<Bytes>) -> io::Result<Self> {
+    pub fn demux(reader: &mut io::Cursor<Bytes>) -> Result<Self, Error> {
         let buf = reader.extract_remaining();
         let mut amf0_reader = Amf0Decoder::new(&buf);
 
-        let name = match amf0_reader.decode_with_type(Amf0Marker::String) {
-            Ok(Amf0Value::String(name)) => name,
-            _ => return Err(io::Error::new(io::ErrorKind::InvalidData, "Invalid script data name")),
+        let Amf0Value::String(name) = amf0_reader.decode_with_type(Amf0Marker::String)? else {
+            unreachable!()
         };
 
-        let data = amf0_reader
-            .decode_all()
-            .map_err(|_| io::Error::new(io::ErrorKind::InvalidData, "Invalid script data"))?;
+        match name.as_ref() {
+            // We might also want to handle "@setDataFrame" the same way as onMetaData.
+            // I'm not sure right now if that is the intended behavior though.
+            "onMetaData" => {
+                let value = amf0_reader.decode()?;
+                println!("{:?}", value);
 
-        Ok(Self {
-            name: name.into_owned(),
-            data: data.into_iter().map(|v| v.to_owned()).collect(),
-        })
+                let Amf0Value::Object(data) = value else { unreachable!() };
+                let data = OnMetaData::try_from(data)?;
+
+                Ok(Self::OnMetaData(Box::new(data)))
+            }
+            "onXMPData" => {
+                let value = amf0_reader.decode()?;
+                println!("{:?}", value);
+
+                let Amf0Value::Object(data) = value else { unreachable!() };
+                let data = OnXmpData::try_from(data)?;
+
+                Ok(Self::OnXmpData(data))
+            }
+            _ => {
+                tracing::warn!(name = %name, "unknown script data");
```

Comment:
i dont know how i feel about tracing here, since we return an other variant the caller can just add this if they want